### ISSUE TITLE
Random spacing issues and error messages in Renderer

### ISSUE THREAD

**Issue by huip-ims** [episphere/quest#438]
@anthonypetersen Hi Tony, We were testing the Module 2 Spanish markdown in the renderer on Friday and today, and we noticed some random spacing issues and error messages popping up.

In the first image below, you can see the following:

- Text boxes now have "Example: …" in gray. However, the 2nd text box has "Example: sum(isDefined".  Is this normal?
- Also, some line breaks are not recognized thus causing random spacing issues which weren't present before (i.e, "O, si le es más fácil recordar en qué año, anótelo aquí:" is now not on a separate line.).
- I'm also seeing random error messages now popping up as shown in red text in the same image, so it looks like some functions that worked previously are now not working. Do you know what may be causing these issues?

In the 2nd image below, it looks like grid row & column headers are now in bold. Also, in some grids, the selection buttons on each row are not aligned. Are these normal and recently added?

Currently, we cannot move forward with our markdown testing in the renderer until these issues are resolved. Could you please help us look into these issues in the renderer?

Thanks,
Pete

![random_issues](https://github.com/episphere/quest/assets/82978291/c87b0647-3a3c-414d-826f-02c9005ec60a)

![Grid_bold_column_row_headers](https://github.com/episphere/quest/assets/82978291/b52c2fd7-96f8-4394-ae29-2468777a500e)

**Comment by huip-ims** [comment]
@anthonypetersen Tony - attached is the Spanish Module 2 markdown in testing.
[module2_ES.txt](https://github.com/episphere/quest/files/15391439/module2_ES.txt)

**Comment by danielruss** [comment]
<img width="616" alt="image" src="https://github.com/episphere/quest/assets/3767675/31f5b4a4-2d9f-4312-a546-905eed618ec7">

This table (though not the same) looks good in chrome, does it look the same for you?  What browser are you using?

**Comment by huip-ims** [comment]
@danielruss @anthonypetersen Daniel - We're using Chrome as well. In your screenshot, although the selection buttons on each row are aligned, the column headers are overwriting each other. In ours, the column headers display correctly, but the selection buttons are not.


![image](https://github.com/episphere/quest/assets/82978291/2a56fc22-ab7c-4fcb-9d4e-33a4c41d1c90)

**Comment by JoeArmani** [comment]
@danielruss I agree; it's related to the table view. I have not seen the misaligned vertical spacing, but Aileen and I found an issue with the rendering of this Module 2 table yesterday (this same 11-column table). The click targets aren't all accessible on a narrower screen for a screen reader, either. I'm actively looking into this.

**Comment by danielruss** [comment]
Let me know if you any help from me.

**Comment by JoeArmani** [comment]
Hi @danielruss, I have a PR in to address:
* Responsive sizing for the grids.
* Improved accessible targeting for the grids.
* Grid misalignment in renderer should be fixed.
* Handle placeholder for all `isDefined` min & max numeric values.

I'm unclear on the errors in the renderer. Here's what I found:
I tested the IMS Spanish code on the current branch and backtested it on earlier commits on the main branch with what appearred to be the same results each time (I was trying to see if something had broken in the process).
Tested the current one
This one: https://github.com/episphere/quest/commit/3e0b2bf67f0c214c243990e6ade1f8853d102339
And this one: https://github.com/episphere/quest/commit/d459d254ea8354038667b004f0a7f5eef6fe41c8
•Results appearred to be the same each time with the sum(isDefined) error.
•Then I found the sum(isDefined) function usage in module1, tested it, and it worked well.
•I traced through evaluateCondition() in the renderer, but I’m not familiar with that part of the app, so I’m not sure what we’re looking for. Thanks for any info you can provide.

Just FYI, I’m OOO on a site visit today but I’ll be back tomorrow.
<img width="578" alt="Screenshot 2024-05-21 at 2 50 44 PM" src="https://github.com/episphere/quest/assets/93854858/9ff55095-36a7-48d1-8fb4-35f65d46701e">

**Comment by JoeArmani** [comment]
I was able to load this module2_ES.txt survey into the PWA. Here's a side-by-side logging comparison with additional logs added.

Here's what's happening:
In the PWA, we have context from the previous results object
<img width="599" alt="Screenshot 2024-05-23 at 2 15 13 PM" src="https://github.com/episphere/quest/assets/93854858/a82574ab-9454-46a6-9670-4a1856789598">

We don't have this context in the renderer. When `replaceValue` gets called, it looks to those previous results to swap out the variable for the value
<img width="857" alt="Screenshot 2024-05-23 at 2 18 10 PM" src="https://github.com/episphere/quest/assets/93854858/6bbcb0d7-0bd7-48f7-bb95-c09597cd2804">

Here's a side-by-side comparison of the process (PWA on the left, with context and renderer on the right without context)
<img width="2311" alt="Screenshot 2024-05-23 at 2 12 27 PM" src="https://github.com/episphere/quest/assets/93854858/ec967f2a-566b-419e-8759-5556e021e125">

So, to make this work in the renderer, we could supply default values for those variables that are passed in when context exists.
Here are the context values we set in the PWA:
`firstName` `age` `AGE` ...other values are CIDs for module numbers (not important in the renderer IMO).

**Comment by danielruss** [comment]
evaluateCondition handles runtime calculations that cannot be handled while parsing, because it may depend on values that are not yet entered.  The line you are pointing to is an old fix to a problem where the answer to a question is the actual number 0.  _How many children do you have?  0_.   The zero is falsy and causes problems.    

This whole section of code needs to be re-written.  Somewhere we started using data-* (think min/max) and we had a mix of data-* and just *.  Massive confusion.

**Comment by JoeArmani** [comment]
Thanks for the heads up on settings. I added that and the validation seems to do what's expected. Am I missing something here or is it possible the `{"AGE":"45"}` setting wasn't added when this was being tested?

Screenshots from renderer in my current working branch...

<img width="1093" alt="Screenshot 2024-05-23 at 4 26 02 PM" src="https://github.com/episphere/quest/assets/93854858/f25d7eac-e9ff-4206-9724-d1ca94247220">
<img width="593" alt="Screenshot 2024-05-23 at 4 24 09 PM" src="https://github.com/episphere/quest/assets/93854858/79351932-a496-43d0-b9e4-992dd7dbcc36">

**Comment by danielruss** [comment]
The development tool does not know about Connect, so it does not get previous values from Firestore.  You need to manually add the values.  However, the previous results you added are stored in indexedDB, so you wont have to enter it again.  (It's actually a mixed blessing).
